README.md
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

![Bolder Flight Systems Logo](img/logo-words_75.png) &nbsp; &nbsp; ![Arduino Logo](img/arduino_logo_75.png)

# Decimal Year
This library converts from year, month, day and hour, minute, second to a decimal year. This library is compatible with Arduino and CMake build systems.
   * [License](LICENSE.md)
   * [Changelog](CHANGELOG.md)
   * [Contributing guide](CONTRIBUTING.md)

# Installation

## Arduino
Use the Arduino Library Manager to install this library or clone to your Arduino/libraries folder. Additionally, the [Bolder Flight Systems Julian Date library](https://github.com/bolderflight/juliandate) must be installed. This library is added as:

```C++
#include "decyear.h"
```

An example Arduino executable is located at *examples/arduino/decyear_ex/decyear_ex.ino*. Teensy 3.x, 4.x, and LC devices are used for testing under Arduino and this library should be compatible with other Arduino devices.

## CMake
CMake is used to build this library, which is exported as a library target called *decyear*. The header is added as:

```C++
#include "decyear.h"
```

The library can be also be compiled stand-alone using the CMake idiom of creating a *build* directory and then, from within that directory issuing:

```
cmake ..
make
```

This will build the library, an example executable called *decyear_example*, and an executable for testing using the Google Test framework, called *decyear_test*. The example executable source files are located at *examples/cmake/decyear_ex.cc*.

# Namespace
This library is within the namespace *bfs*.

# Functions

**double decyear(const int16_t yyyy, const int8_t mm, const int8_t dd, const int8_t h, const int8_t m, const int8_t s)** Given the UTC year, month, day, hour, minute, and second, computes and returns the decimal year.

```C++
/* Decimal year on 2023-5-30 13:23:10 UTC, 2023.40974727296 */
std::cout << std::setprecision(20) << bfs::decyear(2023, 5, 30, 13, 23, 10) << std::endl;
```

**double decyear(const int16_t yyyy, const int8_t mm, const int8_t dd)** Given the UTC year, month, and day, computes and returns the decimal year.

```C++
/* Decimal year on 2023-5-30 0:0:0 UTC, 2023.40821917808 */
Serial.println(bfs::decyear(2023, 5, 30), 20);
```


CHANGELOG.md
# Changelog

## v1.1.0
- Updated CMake to enable local and BFS server builds

## v1.0.0
- Initial baseline


tests/decyear_test.cc
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2023 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/

#include "decyear.h"
#include "gtest/gtest.h"
#include <cmath>

/* Test a random date */
TEST(DecYear, Date) {
  EXPECT_FLOAT_EQ(2023.40974727296, bfs::decyear(2023, 5, 30, 13, 23, 10));
  EXPECT_FLOAT_EQ(2023.40821917808, bfs::decyear(2023, 5, 30));
}

/* Test a leap year */
TEST(DecYear, LeapYear) {
  EXPECT_FLOAT_EQ(2016.41135998533, bfs::decyear(2016, 5, 30, 13, 23, 10));
  EXPECT_FLOAT_EQ(2016.40983606557, bfs::decyear(2016, 5, 30));
}


src/decyear.h
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2023 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/

#ifndef DECYEAR_SRC_DECYEAR_H_  // NOLINT
#define DECYEAR_SRC_DECYEAR_H_

#if defined(ARDUINO)
#include "Arduino.h"
#else
#include <cstddef>
#include <cstdint>
#endif

namespace bfs {

double decyear(const int16_t yyyy, const int8_t mm, const int8_t dd,
               const int8_t h, const int8_t m, const int8_t s);
double decyear(const int16_t yyyy, const int8_t mm, const int8_t dd);

}  // namespace bfs

#endif  // DECYEAR_SRC_DECYEAR_H_ NOLINT


src/decyear.cpp
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2023 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/

#if defined(ARDUINO)
#include "Arduino.h"
#else
#include <cstddef>
#include <cstdint>
#endif

#include "decyear.h"  // NOLINT
#include "juliandate.h"  // NOLINT

namespace bfs {

double decyear(const int16_t yyyy, const int8_t mm, const int8_t dd,
               const int8_t h, const int8_t m, const int8_t s) {
  double jd = juliandate(yyyy, mm, dd, h, m, s);
  double epoch = juliandate(yyyy, 1, 1, 0, 0, 0);
  bool leap_year = false;
  if (yyyy % 400 == 0) {
    leap_year = true;
  } else if (yyyy % 100 == 0) {
    leap_year = false;
  } else if (yyyy % 4 == 0) {
    leap_year = true;
  }
  double decyear;
  if (leap_year) {
    decyear = static_cast<double>(yyyy) + (jd - epoch) / 366.0;
  } else {
    decyear = static_cast<double>(yyyy) + (jd - epoch) / 365.0;
  }
  return decyear;
}

double decyear(const int16_t yyyy, const int8_t mm, const int8_t dd) {
  return decyear(yyyy, mm, dd, 0, 0, 0);
}

}  // namespace bfs


examples/cmake/decyear_ex.cc
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2023 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/

#include "decyear.h"
#include <iostream>
#include <iomanip>

int main() {
  /* Decimal year on 2023-5-30 13:23:10 UTC, 2023.40974727296 */
  std::cout << std::setprecision(20) << bfs::decyear(2023, 5, 30, 13, 23, 10) << std::endl;
  /* Decimal year on 2023-5-30 0:0:0 UTC, 2023.40821917808 */
  std::cout << std::setprecision(20) << bfs::decyear(2023, 5, 30) << std::endl;}
